### PR TITLE
refactor: Refactor all APIs into classes

### PR DESCRIPTION
**Pull request by ghost**
Changes:
- Refactored all APIs into separate classes
- Modified `MetadataMap` to not only take a `AxiosResponse` as parameter
- Improved stability on tests

Most of the changes is moving code from one file to another

### CODE REVIEW

File: src/resources/3d/models3DApi.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  private retrieveEndpoint: Models3DRetrieveEndpoint;
+
+  /** @hidden */
+  constructor(project: string, instance: AxiosInstance, map: MetadataMap) {
+    const path = projectUrl(project) + '/3d/models';
+    this.listEndpoint = generateListEndpoint(instance, path, map, false);
+    this.createEndpoint = generateCreateEndpoint(instance, path, map);
+    this.updateEndpoint = generateUpdateEndpoint(instance, path, map);
+    this.deleteEndpoint = generateDeleteEndpoint(instance, path, map);
+    this.retrieveEndpoint = generateRetrieveSingleEndpoint(instance, path, map);
```
Comment (polomani) [inline src/resources/3d/models3DApi.ts]: discussed this one with you

File: src/resources/apiKeys/apiKeysApi.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  /** @hidden */
+  constructor(project: string, instance: AxiosInstance, map: MetadataMap) {
+    const path = projectUrl(project) + '/apikeys';
+    this.createEndpoint = generateCreateEndpoint(instance, path, map);
+    this.listEndpoint = generateListNoCursorEndpointWithQueryParams(
+      instance,
+      path,
+      map
+    );
+    this.deleteEndpoint = generateDeleteEndpoint(instance, path, map);
```
Comment (polomani) [inline src/resources/apiKeys/apiKeysApi.ts]: same thing, right?

File: src/resources/assets/assetsApi.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      instance,
+      path,
+      map,
+      assetChunker
+    );
+    this.listEndpoint = generateListEndpoint(instance, path, map, true);
+    this.retrieveEndpoint = generateRetrieveEndpoint(instance, path, map);
+    this.updateEndpoint = generateUpdateEndpoint(instance, path, map);
+    this.searchEndpoint = generateSearchEndpoint(instance, path, map);
+    this.deleteEndpoint = generateDeleteEndpoint(instance, path, map);
```
Comment (polomani) [inline src/resources/assets/assetsApi.ts]: same thing, redundant code

File: src/resources/dataPoints/dataPointsApi.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  DatapointsPostDatapoint,
+  LatestDataBeforeRequest,
+} from '../../types/types';
+import { projectUrl } from '../../utils';
+
+export class DataPointsAPI {
+  private insertEndpoint: DataPointsInsertEndpoint;
+  private retrieveEndpoint: DataPointsRetrieveEndpoint;
+  private retrieveLatestEndpoint: DataPointsRetrieveLatestEndpoint;
+  private deleteEndpoint: DataPointsDeleteEndpoint;
```
Comment (polomani) [inline src/resources/dataPoints/dataPointsApi.ts]: same here

File: src/resources/events/eventsApi.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+} from '../../types/types';
+import { projectUrl } from '../../utils';
+
+export class EventsAPI {
+  private createEndpoint: EventsCreateEndpoint;
+  private listEndpoint: EventsListEndpoint;
+  private retrieveEndpoint: EventsRetrieveEndpoint;
+  private updateEndpoint: EventsUpdateEndpoint;
+  private searchEndpoint: EventsSearchEndpoint;
+  private deleteEndpoint: EventsDeleteEndpoint;
```
Comment (polomani) [inline src/resources/events/eventsApi.ts]: and here

File: src/resources/files/filesApi.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+} from './filesUtils';
+
+export class FilesAPI {
+  private listEndpoint: FilesListEndpoint;
+  private uploadEndpoint: FilesUploadEndpoint;
+  private retrieveEndpoint: FilesRetrieveEndpoint;
+  private searchEndpoint: FilesSearchEndpoint;
+  private deleteEndpoint: FilesDeleteEndpoint;
+  private getDownloadUrlsEndpoint: FilesGetDownloadUrlsEndpoint;
+  private updateEndpoint: FilesUpdateEndpoint;
```
Comment (polomani) [inline src/resources/files/filesApi.ts]: and here

File: src/resources/groups/groupsApi.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+} from '../../types/types';
+import { projectUrl } from '../../utils';
+
+export class GroupsAPI {
+  private path: string;
+  private instance: AxiosInstance;
+  private map: MetadataMap;
+  private listEndpoint: GroupsListEndpoint;
+  private createEndpoint: GroupsCreateEndpoint;
+  private deleteEndpoint: GroupsDeleteEndpoint;
```
Comment (polomani) [inline src/resources/groups/groupsApi.ts]: here as well

File: src/resources/projects/projectsApi.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  generateRetrieveSingleEndpoint,
+  generateSingleReplaceEndpoint,
+} from '../../standardMethods';
+import { ProjectResponse, ProjectUpdate } from '../../types/types';
+import { apiUrl, projectUrl } from '../../utils';
+
+export class ProjectsAPI {
+  private instance: AxiosInstance;
+  private map: MetadataMap;
+  private retrieveEndpoint: ProjectsRetrieveEndpoint;
```
Comment (polomani) [inline src/resources/projects/projectsApi.ts]: here

File: src/resources/raw/rawApi.ts [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    this.map = map;
+    const path = (this.path = projectUrl(project) + '/raw/dbs');
+    this.listDatabasesEndpoint = generateListEndpoint(
+      instance,
+      this.path,
+      map,
+      false
+    );
+    this.createDatabasesEndpoint = generateCreateEndpoint(instance, path, map);
+    this.deleteDatabasesEndpoint = generateDeleteEndpoint(instance, path, map);
```
Comment (polomani) [inline src/resources/raw/rawApi.ts]: here

File: src/resources/securityCategories/securityCategoriesApi.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  ListSecurityCategories,
+  SecurityCategory,
+  SecurityCategorySpec,
+} from '../../types/types';
+import { projectUrl } from '../../utils';
+
+export class SecurityCategoriesAPI {
+  private listEndpoint: SecurityCategoriesListEndpoint;
+  private createEndpoint: SecurityCategoriesCreateEndpoint;
+  private deleteEndpoint: SecurityCategoriesDeleteEndpoint;
```
Comment (polomani) [inline src/resources/securityCategories/securityCategoriesApi.ts]: same right here

File: src/resources/serviceAccounts/serviceAccountsApi.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  CogniteInternalId,
+  ServiceAccount,
+  ServiceAccountInput,
+} from '../../types/types';
+import { projectUrl } from '../../utils';
+
+export class ServiceAccountsAPI {
+  private listEndpoint: ServiceAccountsListEndpoint;
+  private createEndpoint: ServiceAccountsCreateEndpoint;
+  private deleteEndpoint: ServiceAccountsDeleteEndpoint;
```
Comment (polomani) [inline src/resources/serviceAccounts/serviceAccountsApi.ts]: here

File: src/resources/timeSeries/timeSeriesApi.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+  /** @hidden */
+  constructor(project: string, instance: AxiosInstance, map: MetadataMap) {
+    const path = projectUrl(project) + '/timeseries';
+    this.createEndpoint = generateCreateEndpoint(instance, path, map);
+    this.listEndpoint = generateListEndpoint(instance, path, map, false);
+    this.searchEndpoint = generateSearchEndpoint(instance, path, map);
+    this.retrieveEndpoint = generateRetrieveEndpoint(instance, path, map);
+    this.updateEndpoint = generateUpdateEndpoint(instance, path, map);
+    this.deleteEndpoint = generateDeleteEndpoint(instance, path, map);
```
Comment (polomani) [inline src/resources/timeSeries/timeSeriesApi.ts]: here

File: src/standardMethods.ts [hunk 13]
```diff
@@ -333,6 +369,9 @@ export function generateSingleReplaceEndpoint<RequestType, RepsonseType>(
       },
       true
     );
-    return metadataMap.addAndReturn(response.data, response);
+    return metadataMap.addAndReturn(
+      response.data,
+      convertAxiosResponseToMetadata(response)
```
Comment (polomani) [inline src/standardMethods.ts]: why do we copy `convertAxiosResponseToMetadata(response)` all the time?
maybe let's just put it inside `addAndReturn`?

**Review by polomani — CHANGES_REQUESTED** [review]
Most of the stuff we discussed orally, but there is one-two small things beside that.

File: src/standardMethods.ts (hunk 13 above)
Comment (ghost) [reply src/standardMethods.ts]: I will simplify it

**Review by polomani — APPROVED** [review]

**Pull request merged**